Issue #34161: Keep loaded global classes reference in GDScriptLanguage

ChibiDenDen opened the pull request:
This change adds a reference to any global class as soon as its loaded.
This removes a case where a global class was loaded more than once.
Specifically fixes an issue where referencing a subclass of a global class does not reference the global class, causing it to be released.
Then, any new reference to the same subclass is different from the original reference, causing problems like #34161.

To be considered:
There might need to be a case where the loaded global classes should be cleared.

ChibiDenDen committed:
Keep loaded global classes reference in GDScriptLanguage

akien-mga commented on modules/gdscript/gdscript.cpp:
```diff
@@ -1945,6 +1945,10 @@ String GDScriptLanguage::get_global_class_name(const String &p_path, String *r_b
 	return String();
 }
 
+void GDScriptLanguage::add_loaded_global_class(Ref<GDScript> global_class) {
```
Function parameters should start with `p_`, so it should be `p_global_class`.

ChibiDenDen committed:
fix argument naming to start with p_

gdscript.cpp GDScriptLanguage
argument name changed is global_class in the add_loaded_global_class

ChibiDenDen committed:
fix argument naming in h file as well

vnen commented:
I'm not fond of this, it essentially keeps everything with a `class_name` loaded forever, and that include its dependencies such as `preload`ed resources. And it only hides a deeper issue, since this goes back as far as #5843.

There must be a way to solve the actual issue that applies to every case.

ChibiDenDen commented:
> I'm not fond of this, it essentially keeps everything with a `class_name` loaded forever, and that include its dependencies such as `preload`ed resources. And it only hides a deeper issue, since this goes back as far as #5843.
> 
> There must be a way to solve the actual issue that applies to every case.

I'm not very familiar with godot source code, but as far as I understand this keeps only GDScript object with a global class name loaded
can you explain what you mean by "everything with a class_name"?

If you have any other idea how to solve it I would be glad to give it a try.

vnen commented:
> I'm not very familiar with godot source code, but as far as I understand this keeps only GDScript object with a global class name loaded
> can you explain what you mean by "everything with a class_name"?

It's what you just said. When you use `class_name` in the script, you're adding it to the global scope (so it will be a "GDScript object with a global class name"). Many people like to give name to all their scripts, so this will keep all scripts loaded. And if those scripts use `preload`, this will also keep those resources alive for the whole application. As you did here, if you never load them it'll be fine, but if you load them once they'll never be unloaded, which isn't desirable.

And this also won't solve the cases when you *don't* have a `class_name` and is using `preload` instead.

> If you have any other idea how to solve it I would be glad to give it a try.

I don't really have one. But I don't like adding this even as workaround, since it's essentially introducing a memory leak for users' projects, even for those who didn't bump into the issue.

vnen reviewed (changes requested):
As I commented, keeping scripts loaded forever is not a good solution since it's essentially a memory leak on purpose.

ChibiDenDen commented:
Closing this PR as a better attempt that does not keep any unused classes in memory exists at:
https://github.com/godotengine/godot/pull/35102
